### PR TITLE
Add newrelic::agent::java class

### PR DESCRIPTION
**Pull request by davehewy**
Hi all,

Added an agent class for java. 

Extremely simple in nature - deals with the fetching of the artifact @ specifed version and unpacking to install_dir with correct perms.

Does not concern itself with the JVM_OPTS etc required to invoke said newrelic.jar with a JVM. 

Raising early for comments and considerations.

```
class{'::newrelic::agent::java':
    license_key => 'somelicensekey'
}
```
pdk tests -> spec/classes/newrelic_java_spec.rb
```
pdk test unit
```
Quite a few failures in the tests but not related to this new class.

### CODE REVIEW

**Comment by craigwatson** [comment]
Hi Dave,

This is a great addition, thanks!

Regarding the tests, Travis does at least seem to fail on the new class only - the module hasn't been explicitly tested with the PDK, just by using the `rake` harness supplied with the module, can you re-test using `bundle exec rake test` and see if you can replicate the errors seen on Travis?

Also, there are a few lint-related style points, all relatively minor - I'll comment on those in-line.

Thanks,
Craig

File: manifests/agent/java.pp [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  String  $package_version               = '3.47.1',
+  String  $install_dir                   = '/opt',
+  String  $newrelic_application_name     = 'webapp',
+  String  $agent_user                    = 'root',
+  String  $agent_group                   = 'root',
+  Boolean $manage_config                 = true
+){
+
+  # == Package Installation
+  package{'unzip': ensure => installed}
```
Comment (craigwatson) [inline manifests/agent/java.pp]: This should be split over multiple lines

File: manifests/agent/java.pp [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  # to the only useful/dependable place for the java agent.
+  # This can be overriden with a private nexus url for example
+  exec{'wget-newrelic-java-agent':
+    path    => ['/usr/bin', '/usr/sbin'],
+    cwd     => $install_dir,
+    command => "wget ${package_source_url}/${package_version}/newrelic-java-${package_version}.zip",
+    creates => "${install_dir}/newrelic-java-${package_version}.zip",
+    onlyif  => "test ! -f ${install_dir}/newrelic-java-${package_version}.zip"
+  }
+  ~> exec{'unzip-newrelic-java-agent-zip':
```
Comment (craigwatson) [inline manifests/agent/java.pp]: Arrow-style resource relationships are discouraged, ideally we should use explicit `requires` or `before` etc on individual resource blocks

File: manifests/agent/java.pp [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  }
+
+  # == Configuration
+
+  # manage the contents of the newrelic.yml or not
+  # if managed only license_key + app_name + env app_names are overriden/supported by module overrides
+  if $manage_config {
+    file{"${install_dir}/newrelic/newrelic.yml":
+      ensure  => 'file',
+      content => epp('newrelic/java/newrelic.yml.epp', {
```
Comment (craigwatson) [inline manifests/agent/java.pp:99]: It doesn't look like this template is contained in the commit - could you ship a boilerplate template, or add one as a test fixture?

**Review by craigwatson — CHANGES_REQUESTED** [review]

File: manifests/agent/java.pp (hunk 3 above)
Comment (davehewy) [reply manifests/agent/java.pp:99]: my bad

Comment (davehewy) [reply manifests/agent/java.pp:99]: rectified

File: manifests/agent/java.pp (hunk 1 above)
Comment (davehewy) [reply manifests/agent/java.pp]: i think "should" is perhaps more of a "could" here. in any case it doesn't warrant a conversation and is done.

File: manifests/agent/java.pp (hunk 2 above)
Comment (davehewy) [reply manifests/agent/java.pp]: Prefer series notify+refreshonly?
I disagree. arrow-style resource relationships are now heavily used & make complete sense to the implementer. if well understood

Comment (davehewy) [reply manifests/agent/java.pp]: @craigwatson ^^

**Comment by davehewy** [comment]
@craigwatson good morning. i've picked up your comments. please take a look when you get a chance :)

**Comment by craigwatson** [comment]
Looks great, thanks!

**Comment by davehewy** [comment]
@craigwatson thanks for the merge. can we get a release shipped this week?

**Comment by craigwatson** [comment]
@DaveHewy Version 2.4.0 has just been released :)

**Pull request merged**